api/oc_push.c

```diff
@@ -0,0 +1,2555 @@
+  }
+
+  return pushd_rsc_rep;
+}
+
+static void
+get_pushd_rsc(oc_request_t *request, oc_interface_mask_t iface_mask,
+              void *user_data)
+{
+  (void)user_data;
+
+  int result = OC_STATUS_OK;
+  oc_pushd_rsc_rep_t *pushd_rsc_rep = _find_pushd_rsc_rep_by_uri(
+    &request->resource->uri, request->resource->device);
+
+  if (!pushd_rsc_rep) {
+    OC_PUSH_ERR("something wrong, can't find resource representation for "
+                "pushed resource (%s)...",
+                oc_string(request->resource->uri));
+    return;
+  }
+
+  if (pushd_rsc_rep->rep) {
+    oc_rep_begin_root_object();
+    switch (iface_mask) {
+    case OC_IF_BASELINE:
+      oc_process_baseline_interface(request->resource);
+      /* fall through */
+    case OC_IF_R:
+    case OC_IF_RW:
+      _build_rep_payload(&root_map, pushd_rsc_rep->rep);
+      break;
+    default:
+      break;
+    }
+    oc_rep_end_root_object();
+
+    oc_send_response(request, result);
+  } else {
+    OC_PUSH_ERR("resource representation for pushed resource (%s) is found, "
+                "but no resource representation for it is built yet!",
+                oc_string(request->resource->uri));
+
+    oc_send_response(request, OC_STATUS_NOT_FOUND);
+  }
+
+  return;
+}
+
+/*
+ * @brief				check if "rt" of pushed resource is part of "rts" (all value of
+ * "rt" should be part of "rts")
+ *
+ * @param recv_obj
+ * @param rep
+ * @return	not 0: found
+ * 			0: not found
+ */
+static bool
+_check_pushd_rsc_rt(oc_recv_t *recv_obj, oc_rep_t *rep)
+{
+  bool result = 0;
+  int rt_len, rts_len;
+  int i, j;
+
+  if (!recv_obj || !rep)
+    return result;
+
+  rts_len = oc_string_array_get_allocated_size(recv_obj->rts);
+
+  /* if "rts" is not configured (""), any pushed resource can be accepted... */
+  if ((rts_len == 1) && !strcmp(oc_string_array_get_item(recv_obj->rts, 0), ""))
+    return 1;
+
+  while (rep) {
+    if ((rep->type == OC_REP_STRING_ARRAY) &&
+        !strcmp(oc_string(rep->name), "rt")) {
+      rt_len = oc_string_array_get_allocated_size(rep->value.array);
+      for (i = 0; i < rt_len; i++) {
+        for (j = 0; j < rts_len; j++) {
+          if (!strcmp(oc_string_array_get_item(rep->value.array, i),
+                      oc_string_array_get_item(recv_obj->rts, j)))
+            break;
+        }
+        if (j == rts_len) {
+          break;
+        }
+      }
+      if (i == rt_len)
+        result = 1;
+
+      break;
+    }
+    rep = rep->next;
+  }
+
+  return result;
+}
+
+static oc_recvs_t *
+_find_recvs_by_device(size_t device_index)
+{
+  oc_recvs_t *recvs_instance = (oc_recvs_t *)oc_list_head(recvs_list);
+
+  while (recvs_instance) {
+    if (recvs_instance->resource->device == device_index) {
+      break;
+    } else {
+      recvs_instance = recvs_instance->next;
+    }
+  }
+
+  return recvs_instance;
+}
+
+/*
+ * @brief				oc_rep_set_pool() should be called before calling this func
+ *
+ * @param new_rep
+ * @param org_rep
+ */
+static void *
```

Comment:
`static oc_rep_t*`, doesn't seem that you need the indeterminate void-ptr type

Reply:
agree, fixed